File: ch4/ruby/heap_sort.rb
class HeapSort
  def initialize(array)
    @array = array
  end

  def sort
    (@array.size / 2 - 1).downto(0) {|i| heapify(i, @array.size) }

    (@array.size - 1).downto(1) {|i|
      @array[0], @array[i] = @array[i], @array[0]
      heapify(0, i)
    }
    @array
  end

  private

  def heapify(idx, max)
    largest = idx
    left = idx * 2 + 1
    right = idx * 2 + 2

    largest = left if left < max && @array[left] > @array[idx]
    largest = right if right < max && @array[right] > @array[largest]
    if largest != idx
      @array[idx], @array[largest] = @array[largest], @array[idx]
      heapify(largest, max)
    end
  end
end

def HeapSort(array)
  HeapSort.new(array).sort
end

File: ch4/cpp/heap_sort.cpp
#include <algorithm>

#include "heap_sort.h"

namespace {

void heapify(int* array, int idx, int max) {
    int largest = idx;
    int left = 2 * idx + 1;
    int right = 2 * idx + 2;

    if (left < max && array[left] > array[idx]) { largest = left; }
    if (right < max && array[right] > array[largest]) { largest = right; }

    if (largest != idx) {
        std::swap(array[idx], array[largest]);
        heapify(array, largest, max);
    }
}

void build_heap(int* array, int size) {
    for (int i = size / 2 + 1; i >= 0; --i) {
        heapify(array, i, size);
    }
}

} // namespace

void heap_sort(int* array, int size) {
    build_heap(array, size);

    for (int i = size - 1; i >= 1; --i) {
        std::swap(array[0], array[i]);
        heapify(array, 0, i);
    }
}
